#28755: [experimental] Proper error handling for checkpoint conflict

dattatrayakumbhar opened the pull request:
"fixes #28755"

**- What I did**
Currently if we created checkpoint which is being used by one container. it does not give proper error.
Provide a proper error if checkpoint is used by container.
Also if we created checkpoint in a custom checkpoint directory then we can't specify checkpoints per container as one container lists all in given custom checkpoint directory.
With this PR, we can now differentiate multiple checkpoint for multiple container in current director.
 
**- How I did it**
1. Wrote a function to check if checkpoint is being used by current container. and give a proper error if new one is already is being used.
2. With custom checkpoint directory, we were not differentiating as per container.
  With this PR, we added <checkpointDir>containerID/checkpoint/checkpointName
  So we could differentiate now.
3. Also remove redundancy and written a function to check directory before create, list, remove.

**- How to verify it**
Create two containers.
docker run -it -d --name=c1 busybox sh
docker run -it -d --name=c2 busybox sh 

1. Create checkpoint without directory
docker checkpoint create c1 check1-wo-c1
                  #should create successfully.
docker checkpoint create c1 check1-wo-c1
                   #Should give error "checkpoint with name check1-wo-c1 already exists for container /c1"
 
2. Create checkpoint with directory 
docker checkpoint create --checkpoint-dir /checkpoint-dir c1 check1-wi-c1
                 #Should get created successfully.
docker checkpoint create --checkpoint-dir /checkpoint-dir c1 check1-wi-c1
                 #checkpoint with name check1-wi-c1 already exists for container /c1

3. Create checkpoint with same as c1 container's checkpoint for c2 in same direcotry.
docker checkpoint create --checkpoint-dir /checkpoint-dir c2 check1-wi-c1
                #Should get create successfully 

4. Create new checkpoint with not a same as c1 container's checkpoint for c2 in same direcotry.
docker checkpoint create --checkpoint-dir /checkpoint-dir c2 check1-wi-c2
                #Should get created successfully. 

5. List down checkpoints with directory.
docker checkpoint ls --checkpoint-dir /checkpoint-dir c1
                 #should list down only checkpoint in given checkpoint directory for c1.
                 CHECKPOINT NAME
                  check1-wi-c1

6. 4. List down checkpoints without directory. 
docker checkpoint ls  c1
                 #Should list checkpoint in default directory only.
                 CHECKPOINT NAME
                check1-wo-c1

7. Delete checkpoint.
docker checkpoint rm --checkpoint-dir /checkpoint-dir c2 check1-wi-c2
                  #Should get deleted successfully.
8. If we listed checkpoint from same directory after deletion, it should show only remained checkpoints.
docker checkpoint ls --checkpoint-dir /checkpoint-dir c2
              CHECKPOINT NAME
               check1-wi-c1

9. #Deletion of deleted/not existing checkpoints should show proper error.
checkpoint rm --checkpoint-dir /checkpoint-dir c2 check1-wi-c2
              #Should give error as "checkpoint check1-wi-c2 is not exist container c2"

![image](https://cloud.githubusercontent.com/assets/21226676/20859366/11329ba6-b983-11e6-810c-d98aa3a8000a.png)


Signed-off-by: dattatrayakumbhar04 <dattatraya.kumbhar@gslab.com>

Commit:
28755: Proper error handling for checkpoint conflict

Signed-off-by: Dattatrayakumbhar <dattatraya.kumbhar@gslab.com>

AkihiroSuda commented on daemon/checkpoint.go:
```diff
@@ -14,6 +14,7 @@ import (
 var (
 	validCheckpointNameChars   = utils.RestrictedNameChars
 	validCheckpointNamePattern = utils.RestrictedNamePattern
+	containerDir               = "/var/lib/docker/containers/"
```
Please avoid hard-coding and use `daemon.repository` instead

AkihiroSuda commented on daemon/checkpoint.go:
```diff
@@ -38,6 +39,25 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
+	var checkpointDirs string
+	containerList := []string{}
+
+	dir, _ := ioutil.ReadDir(containerDir)
```
Let's catch all the errors

AkihiroSuda commented on daemon/checkpoint.go:
```diff
@@ -38,6 +39,25 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
+		checkpoints, _ := ioutil.ReadDir(checkpointDirs)
+
+		for _, checkpoint := range checkpoints {
+			if checkpoint.IsDir() && (checkpoint.Name() == config.CheckpointID) {
```
you don't need parenthesis for the second comparison expression

AkihiroSuda commented on daemon/checkpoint.go:
```diff
@@ -38,6 +39,25 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
+		containerList = append(containerList, d.Name())
+	}
+
+	for _, container := range containerList {
```
why do you need this heavy iteration?
since the error is ` mkdir /var/lib/docker/containers/c4ba0c844c93a31db50dcda57fc0126fd17ec1c4a73cd550a501aac117685a6c/checkpoints/checkpoint2:
file exists`, you just consider the existence of `filepath.Join(daemon.repository, container.ID, "checkpoints", config.checkpointID)`

AkihiroSuda reviewed (changes requested):
As commented, I think the design needs to be refined.

Also, can you add an integration test that fails without this PR but succeeds with this PR?
(Note that current code does not include any IT for checkpoint yet..)

dattatrayakumbhar04 commented:
@AkihiroSuda ptal.
Updated PR with changes for duplicate checkpoint.
Will be adding integration test cases.

milindchawre commented:
LGTM

thaJeztah commented on daemon/checkpoint.go:
```diff
@@ -38,6 +38,11 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
+	checkDir := filepath.Join(daemon.repository, container.ID, "checkpoints", config.CheckpointID)
+	if stat, err := os.Stat(checkDir); err == nil && stat.IsDir() {
+		return fmt.Errorf(" checkpoint with name %s already exists for container %s", config.CheckpointID, container.Name)
+	}
```
This doesn't look correct, because a custom `--checkpoint-dir` can be specified, and this will not catch that.

This should probably be something like this;

```patch
diff --git a/daemon/checkpoint.go b/daemon/checkpoint.go
index 33c6001..87006ea 100644
--- a/daemon/checkpoint.go
+++ b/daemon/checkpoint.go
@@ -38,9 +38,8 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
                return fmt.Errorf("Invalid checkpoint ID (%s), only %s are allowed", config.CheckpointID, validCheckpointNameChars)
        }

-       checkDir := filepath.Join(daemon.repository, container.ID, "checkpoints", config.CheckpointID)
-       if stat, err := os.Stat(checkDir); err == nil && stat.IsDir() {
-               return fmt.Errorf(" checkpoint with name %s already exists for container %s", config.CheckpointID, container.Name)
+       if stat, err := os.Stat(filepath.Join(checkpointDir, config.CheckpointID)); err == nil && stat.IsDir() {
+               return fmt.Errorf("Checkpoint with name %s already exists for container %s", config.CheckpointID, container.Name)
        }

        err = daemon.containerd.CreateCheckpoint(container.ID, config.CheckpointID, checkpointDir, config.Exit)
```

thaJeztah commented on daemon/checkpoint.go:
```diff
@@ -38,6 +38,11 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
+	}
+
 	err = daemon.containerd.CreateCheckpoint(container.ID, config.CheckpointID, checkpointDir, config.Exit)
 	if err != nil {
```
Ideally, we should probably catch the error here, and make a nicer message, however, I'm not sure what the correct way is to get the error-code, and if they are defined;

```
ERRO[0014] Handler for POST /v1.26/containers/test/checkpoints returned error: Cannot checkpoint container test: rpc error: code = 6 desc = mkdir /var/lib/docker/containers/86292a6c7d41387bf1a95a2981e

ERRO[0311] Handler for POST /v1.26/containers/test2/checkpoints returned error: Cannot checkpoint container test2: rpc error: code = 2 desc = exit status 1: "criu failed: type NOTIFY errno 0\nlog file.....
```

dattatrayakumbhar replied:
@thaJeztah,
Thanks for the review, I will update the PR accordingly.

I observed that when I created checkpoint with custom directory(--checkpoint-dir ) option,  "docker checkpoint ls" command does not show any output.

When checkpoint is create without custom checkpoint-dir, checkpoint is created under /var/lib/docker/container/<containerID>/checkpoints/
 
How can we find dependancy between checkpoints and containers when multiple checkpoints created for different containers in same directory using --checkpoint-dir ?

thaJeztah replied:
@dattatrayakumbhar yes, I think that's something that needs to be improved, or we should remove the functionality of a custom checkpoint-dir (and replace with, e.g., an option so save the checkpoint data to a tar, and to load from that tar).

Checkpoint/restore is still experimental, and UX needs feedback (and improvement)

vdemeester commented:
ping @dattatrayakumbhar could you take care of @thaJeztah comments ?

dattatrayakumbhar commented:
yes will start working on it ..

dattatrayakumbhar commented:
Updated PR with changes.
Added testing instructions in description section.

Following changes are  done as part of this PR:
1. if created checkpoint with same name in default directory, it will show proper error that checkpoint with name is already exists.
2. If we created multiple checkpoints with same name in custom checkpointDir.
   **Before this PR**:  It was creating all checkpoints in given checkpoint directory only. There was no mapping of container and checkpoint, so if created same named checkpoint  as container1 for container2.
  **This PR**:  Added support so that multiple checkpoints with same can be created in one directory as (custom-checkpointDir/containerID/checkpoints/<checkpointName>) for different container.
 If we created same name of checkpoint for container1 with checkpointDir. It gives proper error.
 If we created checkpoint for container2 in same checkpointDir and used same checkpoint name for container1 then it gets created successfully for container2 also.

3.  List down checkpoint in customcheckpointDir.
  **Before this PR**: If we created checkpoint for container1 and container2 in same directory, it was listing down all checkpoints for each container.
 **This PR** :  It lists down right checkpoints for right container with same checkpointDir only.
  if checkpoint1 is for container1 in dir1 and checkpoint2 is for container2 in dir1 only then container1 shows checkpoint1 only.

4. Similarly as #3,  It also deletes right checkpoint for right container.
5. Added error handling if checkpoint is not exist in default as well as custom checkpointDir.

@thaJeztah @vdemeester @AkihiroSuda 
Please review this.  This is simple way and same way as default checkpointDir to store mapping of container asd checkpoint.

mlaventure commented on daemon/checkpoint.go:
```diff
@@ -56,13 +62,20 @@ func (daemon *Daemon) CheckpointDelete(name string, config types.CheckpointDelet
-	return os.RemoveAll(filepath.Join(checkpointDir, config.CheckpointID))
+	checkpointAbsDir = filepath.Join(checkpointDir, config.CheckpointID)
+	if stat, err := os.Stat(checkpointAbsDir); err == nil && stat.IsDir() {
+		out = os.RemoveAll(filepath.Join(checkpointDir, config.CheckpointID))
```
you already have the abs path in `checkpointAbsDir` might as well use it.

mlaventure commented on daemon/checkpoint.go:
```diff
@@ -56,13 +62,20 @@ func (daemon *Daemon) CheckpointDelete(name string, config types.CheckpointDelet
+	if stat, err := os.Stat(checkpointAbsDir); err == nil && stat.IsDir() {
+		out = os.RemoveAll(filepath.Join(checkpointDir, config.CheckpointID))
+	} else {
+		out = fmt.Errorf("checkpoint %s does not exist for container %s", config.CheckpointID, name)
```
We may just get a permission denied here, in which case we want to print the error.

mlaventure commented on daemon/checkpoint.go:
```diff
@@ -28,16 +28,22 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
 
 	var checkpointDir string
+	var checkpointAbsDir string
 	if config.CheckpointDir != "" {
```
Since this is repeated 3 times, might as well put it into a small routine.

Since there's also always a `os.Stat` following, What about something like this:
```
func getCheckpointDir(dir string, new bool) (base string, abs string, err error)
```

`new` is here to distinguish call from create and the other operations. if `new` is set then the routine should return an error if  `os.Stat` succeeds (whether the target is a directory or not), otherwise it should return an error is the target doesn't exist, is not a directory or any other system error.

mlaventure reviewed (changes requested):
Also would you mind changing the string in `fmt.Errorf` to start with a lowercase?

dattatrayakumbhar commented:
@mlaventure
Could you review this PR,  I have updated it with changes.

mlaventure commented on daemon/checkpoint.go:
```diff
@@ -16,6 +16,33 @@ var (
 )
 
+// getCheckpointDir verifies checkpoint directory for create,remove, list options and checks if checkpoint already exists
+func getCheckpointDir(checkDir, checkpointID string, name string, containerID string, containercheckpointDir string, new bool) (checkpointDir string, err error) {
```
This is missing a few checks and can be simplified. what about:
```go
func getCheckpointDir(rootDir, id, containerName, containerID, defaultRootDir string, create bool) (checkpointDir string, err error) {
	if rootDir != "" {
		checkpointDir = filepath.Join(rootDir, containerID, "checkpoints")
	} else {
		checkpointDir = defaultRootDir
	}

	checkpoint := filepath.Join(checkpointDir, id)
	stat, err := os.Stat(checkpoint)
	if create {
		switch {
		case err != nil && os.IsNotExist(err):
			// all is good, we can create a new checkpoint with that id in `checkpointDir`
			err = nil
		case err != nil:
			// just return that error (e.g. permission denied)
		case err == nil && stat.IsDir():
			err = fmt.Errorf("a checkpoint with id %s already exists for container %s", id, containerName)
		case err == nil:
			err = fmt.Errorf("%s exist and is not a directory", checkpoint)
		}

		return
	}

	switch {
	case err != nil:
		// something is wrong
	case err == nil && stat.IsDir():
		// this looks like a valid checkpoint
	case err == nil:
		err = fmt.Errorf("%s exist and is not a directory", checkpoint)
	}

	return
}
```

?

mlaventure commented on daemon/checkpoint.go:
```diff
@@ -27,17 +54,16 @@ func (daemon *Daemon) CheckpointCreate(name string, config types.CheckpointCreat
+	checkpointDir, err := getCheckpointDir(config.CheckpointDir, config.CheckpointID, name, container.ID, container.CheckpointDir(), true)
+
+	if err != nil {
+		return fmt.Errorf("Cannot checkpoint container %s: %s", name, err)
```
we usually start errors strings with a lowecase letter.

dattatrayakumbhar commented:
@mlaventure @thaJeztah 
Could you please review this, I have implemented changes?

mlaventure commented:
@dattatrayakumbhar could you change the commit message to reflect the change instead of the issue being fixed?

Something like "Properly report checkpoint directory conflicts" or similar, maybe @thaJeztah has a better phrasing.

ping @thaJeztah @AkihiroSuda for review

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "28755_Err_Duplicate_checkpoint" git@github.com:dattatrayakumbhar/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

mlaventure commented:
@dattatrayakumbhar You need a space between your first name and your last name normally :)

mlaventure reviewed (approved):
LGTM

dattatrayakumbhar commented:
ping @thaJeztah @mlaventure to review this PR if needed to add more detail.

thaJeztah commented on daemon/checkpoint.go:
```diff
@@ -16,6 +16,41 @@ var (
 )
 
+// getCheckpointDir verifies checkpoint directory for create,remove, list options and checks if checkpoint already exists
+func getCheckpointDir(checkDir, checkpointID string, name string, containerID string, containercheckpointDir string, create bool) (checkpointDir string, err error) {
```
Looks like the output variables aren't really used; this could be

```go
func getCheckpointDir(checkDir, checkpointID string, name string, containerID string, containercheckpointDir string, create bool) (string, error) { 
	var checkpointDir string
	var errorVar error
```

In addition, I would;

- change `errorVar` to a different name (`Var` doesn't add much value); perhaps simply `err2`
- change the `name` argument to something that explains it's the _container_'s name, not the checkpoint
- `containercheckpointDir` should use camelcase, `containerCheckpointDir`

(perhaps use `ctrName`, `ctrID`, `ctrCheckpointDir`)


I've looked at possibly use not that many arguments, and instead something like;

```go
func getCheckpointDir(ctr *container.Container, checkpoint types.CheckpointOptions, create bool) (string, error) {
```

Which would make it a lot cleaner, but that's a bigger change, so not something for this PR

thaJeztah commented on daemon/checkpoint.go:
```diff
@@ -16,6 +16,41 @@ var (
+		case err != nil:
+			errorVar = err
+		case err == nil:
+			errorVar = fmt.Errorf("%s exist and is not a directory", checkpointAbsDir)
```
s/exist/exists/

thaJeztah reviewed (changes requested):
Thanks @dattatrayakumbhar, and sorry for being so late reviewing. I tested this, and it improves things (I found some other issues while testing, but will open separate issues for that)

I left some comments inline 👍

dattatrayakumbhar commented:
@thaJeztah,
Thanks for reviewing it.
I have addressed the all comments. Could you please review it again ?

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
Looks like @AkihiroSuda's comments were already addressed, so let me go ahead and merge! :tada:
